src/dictionary.c

```diff
@@ -0,0 +1,163 @@
+#include "dictionary.h"
+#include "redismodule.h"
+
+Trie *SpellCheck_OpenDict(RedisModuleCtx *ctx, const char *dictName, int mode,
+                                 RedisModuleKey **k) {
+  RedisModuleString *keyName = RedisModule_CreateStringPrintf(ctx, DICT_KEY_FMT, dictName);
+
+  *k = RedisModule_OpenKey(ctx, keyName, mode);
+
+  RedisModule_FreeString(ctx, keyName);
+
+  int type = RedisModule_KeyType(*k);
+  if (type == REDISMODULE_KEYTYPE_EMPTY) {
+    Trie *t = NULL;
+    if (mode == REDISMODULE_WRITE) {
+      t = NewTrie();
+      RedisModule_ModuleTypeSetValue(*k, TrieType, t);
+    } else {
+      RedisModule_CloseKey(*k);
+    }
+    return t;
+  }
+
+  if (type != REDISMODULE_KEYTYPE_MODULE || RedisModule_ModuleTypeGetType(*k) != TrieType) {
+    RedisModule_CloseKey(*k);
+    return NULL;
+  }
+
+  return RedisModule_ModuleTypeGetValue(*k);
+}
+
+int Dictionary_Add(RedisModuleCtx *ctx, const char *dictName, RedisModuleString **values,
+                       int len, char **err) {
+  int valuesAdded = 0;
+  RedisModuleKey *k = NULL;
+  Trie *t = SpellCheck_OpenDict(ctx, dictName, REDISMODULE_WRITE, &k);
+  if (t == NULL) {
+    *err = "could not open dict key";
+    return -1;
+  }
+
+  for (int i = 0; i < len; ++i) {
+    valuesAdded += Trie_Insert(t, values[i], 1, 1, NULL);
+  }
+
+  RedisModule_CloseKey(k);
+
+  return valuesAdded;
+}
+
+int Dictionary_Del(RedisModuleCtx *ctx, const char *dictName, RedisModuleString **values,
+                       int len, char **err) {
+  int valuesDeleted = 0;
+  RedisModuleKey *k = NULL;
+  Trie *t = SpellCheck_OpenDict(ctx, dictName, REDISMODULE_WRITE, &k);
+  if (t == NULL) {
+    *err = "could not open dict key";
+    return -1;
+  }
+
+  for (int i = 0; i < len; ++i) {
+    size_t len;
+    const char *val = RedisModule_StringPtrLen(values[i], &len);
+    valuesDeleted += Trie_Delete(t, (char *)val, len);
```

Comment:
We should change trie to accept `const char*` and not `char*` (not a blocker for this review, just something that's annoying).